subsys/net/lib/sntp/sntp.c

```diff
@@ -0,0 +1,261 @@
+		NET_DBG("orig_tm_f:       %x", pkt->orig_tm_f);
+		NET_DBG("rx_tm_s:         %x", pkt->rx_tm_s);
+		NET_DBG("rx_tm_f:         %x", pkt->rx_tm_f);
+		NET_DBG("tx_tm_s:         %x", pkt->tx_tm_s);
+		NET_DBG("tx_tm_f:         %x", pkt->tx_tm_f);
+	}
+}
+
+static s32_t parse_response(u8_t *data, u16_t len, u32_t orig_ts,
+				u64_t *epoch_time)
+{
+	struct sntp_pkt *pkt = (struct sntp_pkt *)data;
+
+	sntp_pkt_dump(pkt);
+
+	if (ntohl(pkt->orig_tm_s) != orig_ts) {
+		NET_DBG("Unexpected originate timestamp");
+		return -EINVAL;
+	}
+
+	if (pkt->li > SNTP_LI_MAX) {
+		NET_DBG("Unexpected LI");
+		return -EINVAL;
+	}
+
+	if (pkt->mode != SNTP_MODE_SERVER) {
+		/* For unicast and manycast, server should return 4.
+		 * For broadcast (which is not supported now), server should
+		 * return 5.
+		 */
+		NET_DBG("Unexpected mode");
+		return -EINVAL;
+	}
+
+	if (pkt->stratum == SNTP_STRATUM_KOD) {
+		NET_DBG("kiss-o'-death stratum");
+		return -EBUSY;
+	}
+
+	if (epoch_time) {
+		u64_t ts = (u64_t)ntohl(pkt->tx_tm_s);
+
+		/* Check if most significant bit is set */
+		if ((ts >> 31) != 0) {
+			/* UTC time is reckoned from 0h 0m 0s UTC
+			 * on 1 January 1900.
+			 */
+			if (ts >= OFFSET_1970_JAN_1) {
+				*epoch_time = ts - OFFSET_1970_JAN_1;
+			} else {
+				return -EINVAL;
+			}
+		} else {
+			/* UTC time is reckoned from 6h 28m 16s UTC
+			 * on 7 February 2036.
+			 */
+			*epoch_time = ts + 0x100000000 - OFFSET_1970_JAN_1;
+		}
+	}
+
+	return 0;
+}
+
+static void sntp_recv_cb(struct net_app_ctx *ctx, struct net_pkt *pkt,
+			int status, void *user_data)
+{
+	struct sntp_ctx *sntp = NULL;
+
+	if (!pkt) {
+		return;
+	}
+
+	if (status < 0) {
+		return;
+	}
+
+	if (!user_data) {
+		goto error_exit;
+	}
+	sntp = (struct sntp_ctx *)user_data;
+
+	if (net_pkt_appdatalen(pkt) != sizeof(struct sntp_pkt)) {
+		goto error_exit;
+	}
+
+	if (!net_pkt_appdata(pkt)) {
+		goto error_exit;
+	}
+
+	k_fifo_put(&sntp->recv_q, pkt);
+	return;
+
+error_exit:
+	if (ctx) {
+		k_fifo_cancel_wait(&sntp->recv_q);
+	}
+	net_pkt_unref(pkt);
+}
+
+static u32_t get_uptime_in_sec(void)
+{
+	u64_t time;
+
+	k_enable_sys_clock_always_on();
+	time = k_uptime_get_32();
+	k_disable_sys_clock_always_on();
+
+	return time / 1000;
+}
+
+int sntp_init(struct sntp_ctx *ctx, const char *srv_addr, u16_t srv_port,
+		u32_t timeout)
+{
+	int rv;
+
+	if (!ctx) {
+		return -EFAULT;
+	}
+
+	memset(ctx, 0, sizeof(struct sntp_ctx));
+
+	k_fifo_init(&ctx->recv_q);
+
+	rv = net_app_init_udp_client(&ctx->net_app_ctx, NULL, NULL, srv_addr,
+					srv_port, ctx->timeout, ctx);
+	if (rv < 0) {
+		NET_DBG("Failed to init udp client");
+		return rv;
+	}
+
+	rv = net_app_set_cb(&ctx->net_app_ctx, NULL, sntp_recv_cb, NULL, NULL);
+	if (rv < 0) {
+		NET_DBG("Failed to set net app callback");
+		return rv;
+	}
+
+	ctx->timeout = timeout;
+	ctx->is_init = true;
+
+	return 0;
+}
+
+int sntp_request(struct sntp_ctx *ctx, u64_t *epoch_time)
+{
+	struct sntp_pkt tx_pkt = {0};
```

Comment:
Hmm, I thought checkpatch.pl would fail this because there is no spaces around 0.